### PR TITLE
[Backport] [2.x] Update codecs to Apache Lucene 9.12.0

### PR DESCRIPTION
**Pull request by reta**
### Description
Backport of https://github.com/opensearch-project/custom-codecs/pull/198 to `2.x`

### Related Issues
Closes https://github.com/opensearch-project/custom-codecs/issues/197

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/custom-codecs/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
@sarthakaggarwal97 I have to bring https://github.com/opensearch-project/custom-codecs/pull/198 to `2.x` first since 2.18.0 is on Apache Lucene 9.12.0. already (and BWC fails `2.x` branch compilation)

File: src/main/java/org/opensearch/index/codec/customcodecs/backward_codecs/lucene99/QatCompressionMode.java [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param mode The compression mode (QAT_LZ4 or QAT_DEFLATE)
+     * @param compressionLevel The compression level to use.
+     * @param supplier a supplier for QAT acceleration mode.
+     */
+    protected QatCompressionMode(Lucene99QatCodec.Mode mode, int compressionLevel, Supplier<QatZipper.Mode> supplier) {
```
Comment (sarthakaggarwal97) [inline src/main/java/org/opensearch/index/codec/customcodecs/backward_codecs/lucene99/QatCompressionMode.java:68]: I don't think we should duplicate QATCompressionMode. This class should be reusable right? We do it for Zstd related CompressionModes currently.

**Review by sarthakaggarwal97 — APPROVED** [review]

**Comment by sarthakaggarwal97** [comment]
@reta looks good to me. Should we merge first to 2.x (since build is green here)?
@andrross any major comments on this PR before we merge?

**Comment by reta** [comment]
> @reta looks good to me. Should we merge first to 2.x (since build is green here)?

Thanks @sarthakaggarwal97 , yes, we need `2.x` first, thanks!

**Comment by sarthakaggarwal97** [comment]
thanks @reta for the change

**Pull request merged**